// dmdb/src/net/msg/prepare.js
"use strict";

var __extends = this && this.__extends || function() {
    var i = function(t, e) {
        return (i = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, e) {
            t.__proto__ = e;
        } || function(t, e) {
            for (var s in e) e.hasOwnProperty(s) && (t[s] = e[s]);
        })(t, e);
    };
    return function(t, e) {
        function s() {
            this.constructor = t;
        }
        i(t, e), t.prototype = null === e ? Object.create(e) : (s.prototype = e.prototype, 
        new s());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var execute_1 = require("./execute"), msg_1 = require("./msg"), const_1 = require("../../desc/const"), executeRetInfo_1 = require("../../desc/executeRetInfo"), parameter_1 = require("../../desc/parameter"), dbtype_1 = require("../../driver/dbtype"), typeDescriptor_1 = require("../../desc/typeDescriptor"), Prepare = function(c) {
    function t(t, e, s, i, r) {
        var h = c.call(this, t, msg_1.Msg.Ni, e, r) || this;
        return h.ye = s, h.be = i, h;
    }
    return __extends(t, c), t.prototype.te = function() {
        var t = this.nr.at, e = this.ir, s = msg_1.Msg.os;
        s += this.mt.l(s, t.conn_prop_autoCommit ? 1 : 0), s += this.mt.l(s, this.ye ? 1 : 0), 
        s += this.mt.l(s, 0), s += this.mt.l(s, 1), s += this.mt.l(s, t.compatibleOracle() ? 0 : 1), 
        s += this.mt.y(s, this.be), s += this.mt.U(s, e.maxRows <= 0 || t.conn_prop_enRsCache ? const_1.Const.INT64_MAX : e.maxRows), 
        s += this.mt.l(s, t.conn_prop_isBdtaRS ? const_1.Const.RS_BDTA : const_1.Const.RS_ROW), 
        s += this.mt.y(s, 0), s += this.mt.l(s, 0), s += this.mt.l(s, 0), s += this.mt.l(s, 0), 
        s += this.mt.v(s, e.queryTimeout), this.mt.l(s, e.innerExec ? 1 : 0), this.mt.Xn(e.nativeSql, t.getServerEncoding());
    }, t.prototype.ee = function() {
        if (this.ye) return c.prototype.ee.call(this);
        var t = new executeRetInfo_1.ExecuteRetInfo(), e = this.nr.at, s = msg_1.Msg.os;
        t.retSqlType = this.mt.r(s), s += const_1.Const.USINT_SIZE;
        var i = this.mt.oi(s);
        s += const_1.Const.USINT_SIZE;
        var r = this.mt.r(s);
        if (s += const_1.Const.USINT_SIZE, this.mt.n(s), s += const_1.Const.DDWORD_SIZE, 
        e.trxStatus = this.mt.t(s), const_1.Const.ULINT_SIZE, 0 < i) {
            this.ir.serverParams = this.me(i), this.ir.bindParams = new Array(i);
            for (var h = 0; h < i; h++) {
                var n = this.ir.serverParams[h], _ = new parameter_1.Parameter();
                _.type = n.type, _.prec = n.prec, _.scale = n.scale, _.nullable = n.nullable, _.hasDefault = n.hasDefault, 
                _.typeFlag = n.typeFlag, _.lob = n.lob, _.ioType = n.ioType, _.name = n.name, _.typeName = n.typeName, 
                _.tableName = n.tableName, _.schemaName = n.schemaName, _.lobTabId = n.lobTabId, 
                _.lobColId = n.lobColId, _.mask = n.mask, _.typeDescriptor = n.typeDescriptor, this.ir.bindParams[h] = _;
            }
            this.ir.paramCount = i;
        } else this.ir.serverParams = new Array(0), this.ir.bindParams = new Array(0), this.ir.paramCount = 0;
        return this.ir.columns = 0 < r ? this.as(r, t.rsBdta) : new Array(0), t;
    }, t.prototype.me = function(t) {
        for (var e, s, i, r, h = new Array(t), n = 0; n < t; n++) {
            var _ = new parameter_1.Parameter();
            _.type = this.mt.N(), _.prec = this.mt.N(), _.scale = this.mt.N(), _.nullable = 0 != this.mt.N();
            var c = this.mt.hn();
            _.hasDefault = 0 != (c & execute_1.Execute.vs), _.typeFlag = 0 != (c & execute_1.Execute.ds) ? parameter_1.Parameter.TYPE_FLAG_RECOMMEND : parameter_1.Parameter.TYPE_FLAG_EXACT, 
            _.lob = 0 != (c & execute_1.Execute.ps), this.mt.N(), _.ioType = this.mt.hn(), e = this.mt.hn(), 
            s = this.mt.hn(), i = this.mt.hn(), r = this.mt.hn(), _.name = this.mt.xn(e, this.nr.at.getServerEncoding()), 
            _.typeName = this.mt.xn(s, this.nr.at.getServerEncoding()), _.tableName = this.mt.xn(i, this.nr.at.getServerEncoding()), 
            _.schemaName = this.mt.xn(r, this.nr.at.getServerEncoding()), _.lob && (_.lobTabId = this.mt.N(), 
            _.lobColId = this.mt.hn()), _.type === dbtype_1.DBType.DATETIME || _.type === dbtype_1.DBType.DATETIME2 ? 0 != (_.scale & dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK) ? (_.scale = _.scale & ~dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK, 
            _.mask = const_1.Const.MASK_LOCAL_DATETIME) : 0 != (_.scale & dbtype_1.DBType.ORACLE_DATE_SCALE_MASK) && (_.scale = _.scale & ~dbtype_1.DBType.ORACLE_DATE_SCALE_MASK, 
            _.mask = const_1.Const.MASK_ORACLE_DATE) : _.type === dbtype_1.DBType.DECIMAL && _.scale === dbtype_1.DBType.ORACLE_FLOAT_SCALE_MASK ? (_.prec = Math.round(.30103 * _.prec) + 1, 
            _.scale = -1, _.mask = const_1.Const.MASK_ORACLE_FLOAT) : _.type === dbtype_1.DBType.VARCHAR && _.prec === dbtype_1.DBType.BFILE_PREC && _.scale === dbtype_1.DBType.BFILE_SCALE && (_.mask = const_1.Const.MASK_BFILE), 
            h[n] = _;
        }
        for (var o, n = 0; n < t; n++) {
            dbtype_1.DBType.isComplexType(h[n].type, h[n].scale) && ((o = new typeDescriptor_1.TypeDescriptor(this.nr.at)).unpack(this.mt), 
            h[n].typeDescriptor = o);
        }
        return h;
    }, t;
}(execute_1.Execute);

exports.Prepare = Prepare;
